spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,415 @@
+            if k == 1:
+                short_form = sample_toks[0]
+            elif k == 2:
+                short_form = sample_toks[0] + os.sep + sample_toks[1]
+            elif k > 2:
+                short_form = sample_toks[0] + "..." + os.sep + sample_toks[k-1]
+            for idx in group.keys():
+                new_path_list[idx] += short_form + (os.sep if k > 0 else '')
+                del level_idx[idx]
+            recurse_level({idx: toks[k:] for idx, toks in group.iteritems()})
+
+    recurse_level(dict(enumerate(path_list)))
+    return [path[:-1] for path in new_path_list]  # trim final sep
+
+
+class HTMLDelegate(QStyledItemDelegate):
+    """With this delegate, a QListWidgetItem can render HTML.
+
+    Taken from http://stackoverflow.com/a/5443112/2399799
+    """
+    def paint(self, painter, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        style = QApplication.style() if options.widget is None\
+            else options.widget.style()
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+
+        options.text = ""
+        style.drawControl(QStyle.CE_ItemViewItem, options, painter)
+
+        ctx = QAbstractTextDocumentLayout.PaintContext()
+
+        textRect = style.subElementRect(QStyle.SE_ItemViewItemText, options)
+        painter.save()
+        painter.translate(textRect.topLeft())
+        painter.setClipRect(textRect.translated(-textRect.topLeft()))
+        doc.documentLayout().draw(painter, ctx)
+
+        painter.restore()
+
+    def sizeHint(self, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+        doc.setTextWidth(options.rect.width())
+        return QSize(doc.idealWidth(), doc.size().height())
+
+
+class UpDownFilter(QObject):
+    """Use with ``installEventFilter`` to get up/down arrow key press signal.
+    """
+    up_down = Signal(int)
+
+    def eventFilter(self, src, e):
+        if isinstance(e, QKeyEvent) and e.type() == QEvent.KeyPress:
+            if e.key() == Qt.Key_Up:
+                self.up_down.emit(-1)
+            elif e.key() == Qt.Key_Down:
+                self.up_down.emit(+1)
+        return super(UpDownFilter, self).eventFilter(src, e)
+
+
+class HelperToolButton(QToolButton):
+    """ """
+    def __init__(self):
+        QToolButton.__init__(self)
+
+    def setToolTip(self, text):
+        """ """
+        self._tip_text = text
+
+    def toolTip(self):
+        """ """
+        return self._tip_text
+
+    def mousePressEvent(self, event):
+        """ """
+        QToolTip.hideText()
+
+    def mouseReleaseEvent(self, event):
+        """ """
+        QToolTip.showText(self.mapToGlobal(QPoint(0, 0)), self._tip_text)
+
+
+class FileSwitcher(QDialog):
+    close_file = Signal(int)
+    edit_file = Signal(int)
+    edit_line = Signal(int)
```

Comment:
Add a comment to explain what the argument means.
@goanpeca Also is always prefixing signals with `sig_` really a thing now ?